src/txmempool.cpp

```diff
@@ -1055,11 +1055,25 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
     }
 }
 
-bool CTxMemPool::TransactionWithinChainLimit(const uint256& txid, size_t chainLimit) const {
+uint64_t CTxMemPool::CalculateDescendantMaximum(txiter entry) const {
+    // find top parent
+    txiter top = entry;
+    for (;;) {
+        const setEntries& parents = GetMemPoolParents(top);
+        if (parents.size() == 0) break;
+        top = *parents.begin();
+    }
+    return top->GetCountWithDescendants();
+}
+
+void CTxMemPool::GetTransactionAncestry(const uint256& txid, int64_t& ancestors, int64_t& descendants) const {
     LOCK(cs);
     auto it = mapTx.find(txid);
-    return it == mapTx.end() || (it->GetCountWithAncestors() < chainLimit &&
-       it->GetCountWithDescendants() < chainLimit);
+    ancestors = descendants = -1;
+    if (it != mapTx.end()) {
+        ancestors = it->GetCountWithAncestors();
+        descendants = (int64_t)CalculateDescendantMaximum(it);
```

Comment:
Maybe add some bounds-checking asserts here ala `assert(int64_t(nCountWithDescendants) > 0);`

Reply:
I'm going to switch back to unsigned, so no need.